File: .shipmate/hooks/ts/detect-struggling.ts
#!/usr/bin/env bun

/**
 * Struggling Detection Hook (beforeShellExecution)
 *
 * Detects when agent is running same/similar commands repeatedly.
 * Injects guidance to try a different approach.
 */

import { createHash } from 'node:crypto';
import { existsSync, mkdirSync, readFileSync, writeFileSync } from 'node:fs';
import { homedir } from 'node:os';
import { join } from 'node:path';

interface HookInput {
  conversation_id: string;
  hook_event_name: string;
  command: string;
  cwd?: string;
}

interface HookOutput {
  permission: 'allow' | 'deny';
  user_message?: string;
  agent_message?: string;
}

interface CommandHistory {
  commands: string[];
  signatures: Record<string, number>;
  exact: Record<string, number>;
  last_update: number;
}

const HISTORY_FILE = join(homedir(), '.shipmate', 'context', 'command_history.json');
const STRUGGLE_THRESHOLD = 3;
const SIMILAR_THRESHOLD = 5;
const HISTORY_WINDOW = 20;

function getCommandSignature(command: string): string {
  const normalized = command
    .replace(/"[^"]*"/g, '""')
    .replace(/'[^']*'/g, "''")
    .replace(/\d+/g, 'N')
    .replace(/(cd|cat|ls|rm|cp|mv|mkdir)\s+\S+/g, '$1 PATH');
  return normalized.trim().toLowerCase();
}

function getCommandHash(command: string): string {
  return createHash('md5').update(command).digest('hex').slice(0, 12);
}

function loadHistory(conversationId: string): CommandHistory {
  const empty: CommandHistory = { commands: [], signatures: {}, exact: {}, last_update: 0 };

  if (!existsSync(HISTORY_FILE)) return empty;

  try {
    const allHistory = JSON.parse(readFileSync(HISTORY_FILE, 'utf-8'));
    return allHistory[conversationId] || empty;
  } catch {
    return empty;
  }
}

function saveHistory(conversationId: string, history: CommandHistory) {
  const contextDir = join(homedir(), '.shipmate', 'context');
  mkdirSync(contextDir, { recursive: true });

  let allHistory: Record<string, CommandHistory> = {};

  if (existsSync(HISTORY_FILE)) {
    try {
      allHistory = JSON.parse(readFileSync(HISTORY_FILE, 'utf-8'));
    } catch {
      allHistory = {};
    }
  }

  // Clean old entries (> 1 hour)
  const cutoff = Date.now() - 3600000;
  for (const key of Object.keys(allHistory)) {
    if ((allHistory[key].last_update || 0) < cutoff) {
      delete allHistory[key];
    }
  }

  history.last_update = Date.now();
  allHistory[conversationId] = history;

  writeFileSync(HISTORY_FILE, JSON.stringify(allHistory));
}

interface StruggleInfo {
  isStruggling: boolean;
  struggleType: string | null;
  count: number;
  suggestion: string | null;
}

function detectStruggling(command: string, history: CommandHistory): StruggleInfo {
  const result: StruggleInfo = {
    isStruggling: false,
    struggleType: null,
    count: 0,
    suggestion: null,
  };

  const cmdHash = getCommandHash(command);
  const cmdSig = getCommandSignature(command);

  // Check exact command repetition
  const exactCount = history.exact[cmdHash] || 0;
  if (exactCount >= STRUGGLE_THRESHOLD) {
    result.isStruggling = true;
    result.struggleType = 'exact_repeat';
    result.count = exactCount;
    result.suggestion =
      'This exact command has been run multiple times. Consider trying a different approach or checking for underlying issues.';
  }

  // Check similar command pattern
  const sigCount = history.signatures[cmdSig] || 0;
  if (sigCount >= SIMILAR_THRESHOLD) {
    result.isStruggling = true;
    result.struggleType = 'similar_pattern';
    result.count = sigCount;
    result.suggestion =
      'Similar commands have been attempted repeatedly. Step back and reconsider the approach.';
  }

  // Check for package install loops
  if (/\b(npm|yarn|pip|go)\s+(install|add|get)/.test(command)) {
    const pkgAttempts = history.commands
      .slice(-10)
      .filter((c) => /\b(npm|yarn|pip|go)\s+(install|add|get)/.test(c)).length;
    if (pkgAttempts >= 3) {
      result.isStruggling = true;
      result.struggleType = 'dependency_issues';
      result.count = pkgAttempts;
      result.suggestion =
        'Multiple package install attempts detected. Check package.json/requirements.txt for conflicts or try clearing cache.';
    }
  }

  // Check for permission issues
  if (/permission denied/i.test(command) || /sudo/.test(command)) {
    const permAttempts = history.commands
      .slice(-5)
      .filter((c) => /sudo/.test(c) || /chmod/.test(c)).length;
    if (permAttempts >= 2) {
      result.isStruggling = true;
      result.struggleType = 'permission_issues';
      result.count = permAttempts;
      result.suggestion =
        'Permission issues detected. Check file ownership and permissions systematically.';
    }
  }

  return result;
}

function updateHistory(history: CommandHistory, command: string): CommandHistory {
  const cmdHash = getCommandHash(command);
  const cmdSig = getCommandSignature(command);

  history.commands = [...history.commands, command].slice(-HISTORY_WINDOW);
  history.exact[cmdHash] = (history.exact[cmdHash] || 0) + 1;
  history.signatures[cmdSig] = (history.signatures[cmdSig] || 0) + 1;

  return history;
}

function logStruggle(conversationId: string, command: string, struggleInfo: StruggleInfo) {
  const now = new Date();
  const dateStr = now.toISOString().split('T')[0];
  const diaryDir = join(homedir(), '.shipmate', 'diary', 'entries', dateStr);

  try {
    mkdirSync(diaryDir, { recursive: true });

    const id = createHash('md5')
      .update(`${conversationId}${now.toISOString()}`)
      .digest('hex')
      .slice(0, 8);

    const entry = {
      id,
      timestamp: now.toISOString(),
      conversation_id: conversationId,
      type: 'struggling',
      severity: 'medium',
      category: 'execution',
      context: {
        command: command.slice(0, 200),
        struggle_type: struggleInfo.struggleType,
        attempt_count: struggleInfo.count,
      },
      learning: {
        what_happened: `Agent struggled: ${struggleInfo.struggleType}`,
        pattern: struggleInfo.struggleType,
        improvement: 'Detect earlier and try different approach',
      },
    };

    writeFileSync(join(diaryDir, `struggle-${id}.json`), JSON.stringify(entry, null, 2));
  } catch {
    // Silently fail
  }
}

async function main() {
  try {
    const input: HookInput = await Bun.stdin.json();
    const output: HookOutput = { permission: 'allow' };

    if (input.command) {
      let history = loadHistory(input.conversation_id);
      const struggleInfo = detectStruggling(input.command, history);

      history = updateHistory(history, input.command);
      saveHistory(input.conversation_id, history);

      if (struggleInfo.isStruggling) {
        logStruggle(input.conversation_id, input.command, struggleInfo);
        output.agent_message = `[Shipmate] ${struggleInfo.suggestion}`;
      }
    }

    console.log(JSON.stringify(output));
  } catch {
    console.log(JSON.stringify({ permission: 'allow' }));
  }
}

main();

File: .shipmate/hooks/detect-struggling.py
#!/usr/bin/env python3
"""
Repeated Command/Struggling Detection Hook (beforeShellExecution)

Cursor Hook: Detects when the agent is struggling by running the same or
similar commands repeatedly. Injects context to try a different approach.

Input (stdin JSON):
{
  "conversation_id": "string",
  "hook_event_name": "beforeShellExecution",
  "command": "string",
  "cwd": "string"
}

Output (stdout JSON):
{
  "permission": "allow",
  "user_message": "optional guidance",
  "agent_message": "optional agent guidance"
}
"""

import json
import sys
import re
import hashlib
from datetime import datetime
from pathlib import Path

# Command history for this session
HISTORY_FILE = Path.home() / ".shipmate" / "context" / "command_history.json"
STRUGGLE_THRESHOLD = 3  # Same command N times = struggling
SIMILAR_THRESHOLD = 5   # Similar commands N times = struggling
HISTORY_WINDOW = 20     # Only look at last N commands


def get_command_signature(command: str) -> str:
    """Get a normalized signature for a command (ignoring variable parts)."""
    # Remove quotes content
    normalized = re.sub(r'"[^"]*"', '""', command)
    normalized = re.sub(r"'[^']*'", "''", normalized)
    # Remove numbers
    normalized = re.sub(r'\d+', 'N', normalized)
    # Remove paths after common commands
    normalized = re.sub(r'(cd|cat|ls|rm|cp|mv|mkdir)\s+\S+', r'\1 PATH', normalized)
    return normalized.strip().lower()


def get_command_hash(command: str) -> str:
    """Get a hash of the exact command."""
    return hashlib.md5(command.encode()).hexdigest()[:12]


def load_history(conversation_id: str) -> dict:
    """Load command history for this conversation."""
    if not HISTORY_FILE.exists():
        return {"commands": [], "signatures": {}, "exact": {}}

    try:
        with open(HISTORY_FILE, 'r') as f:
            all_history = json.load(f)
        return all_history.get(conversation_id, {"commands": [], "signatures": {}, "exact": {}})
    except Exception:
        return {"commands": [], "signatures": {}, "exact": {}}


def save_history(conversation_id: str, history: dict):
    """Save command history."""
    HISTORY_FILE.parent.mkdir(parents=True, exist_ok=True)

    all_history = {}
    if HISTORY_FILE.exists():
        try:
            with open(HISTORY_FILE, 'r') as f:
                all_history = json.load(f)
        except Exception:
            pass

    # Keep only recent conversations (last hour)
    cutoff = datetime.now().timestamp() - 3600
    all_history = {
        k: v for k, v in all_history.items()
        if v.get("last_update", 0) > cutoff
    }

    history["last_update"] = datetime.now().timestamp()
    all_history[conversation_id] = history

    with open(HISTORY_FILE, 'w') as f:
        json.dump(all_history, f)


def detect_struggling(command: str, history: dict) -> dict:
    """Detect if we're struggling based on command patterns."""
    result = {
        "is_struggling": False,
        "struggle_type": None,
        "count": 0,
        "suggestion": None
    }

    cmd_hash = get_command_hash(command)
    cmd_sig = get_command_signature(command)

    # Check exact command repetition
    exact_count = history.get("exact", {}).get(cmd_hash, 0)
    if exact_count >= STRUGGLE_THRESHOLD:
        result["is_struggling"] = True
        result["struggle_type"] = "exact_repeat"
        result["count"] = exact_count
        result["suggestion"] = "This exact command has been run multiple times. Consider trying a different approach or checking for underlying issues."

    # Check similar command pattern
    sig_count = history.get("signatures", {}).get(cmd_sig, 0)
    if sig_count >= SIMILAR_THRESHOLD:
        result["is_struggling"] = True
        result["struggle_type"] = "similar_pattern"
        result["count"] = sig_count
        result["suggestion"] = "Similar commands have been attempted repeatedly. Step back and reconsider the approach."

    # Check for specific struggle patterns
    if re.search(r'\b(npm|yarn|pip|go)\s+(install|add|get)', command):
        pkg_attempts = sum(1 for c in history.get("commands", [])[-10:]
                         if re.search(r'\b(npm|yarn|pip|go)\s+(install|add|get)', c))
        if pkg_attempts >= 3:
            result["is_struggling"] = True
            result["struggle_type"] = "dependency_issues"
            result["count"] = pkg_attempts
            result["suggestion"] = "Multiple package install attempts detected. Check package.json/requirements.txt for conflicts or try clearing cache."

    # Check for permission issues
    if 'permission denied' in command.lower() or 'sudo' in command:
        perm_attempts = sum(1 for c in history.get("commands", [])[-5:]
                          if 'sudo' in c or 'chmod' in c)
        if perm_attempts >= 2:
            result["is_struggling"] = True
            result["struggle_type"] = "permission_issues"
            result["count"] = perm_attempts
            result["suggestion"] = "Permission issues detected. Check file ownership and permissions systematically."

    return result


def update_history(history: dict, command: str) -> dict:
    """Update history with new command."""
    cmd_hash = get_command_hash(command)
    cmd_sig = get_command_signature(command)

    # Add to command list (keep last N)
    history["commands"] = (history.get("commands", []) + [command])[-HISTORY_WINDOW:]

    # Update exact counts
    if "exact" not in history:
        history["exact"] = {}
    history["exact"][cmd_hash] = history["exact"].get(cmd_hash, 0) + 1

    # Update signature counts
    if "signatures" not in history:
        history["signatures"] = {}
    history["signatures"][cmd_sig] = history["signatures"].get(cmd_sig, 0) + 1

    return history


def log_struggle(conversation_id: str, command: str, struggle_info: dict):
    """Log struggling incident to diary."""
    diary_dir = Path.home() / ".shipmate" / "diary" / "entries" / datetime.now().strftime("%Y-%m-%d")
    diary_dir.mkdir(parents=True, exist_ok=True)

    entry = {
        "id": hashlib.md5(f"{conversation_id}{datetime.now().isoformat()}".encode()).hexdigest()[:8],
        "timestamp": datetime.now().isoformat(),
        "conversation_id": conversation_id,
        "type": "struggling",
        "severity": "medium",
        "category": "execution",
        "context": {
            "command": command[:200],
            "struggle_type": struggle_info["struggle_type"],
            "attempt_count": struggle_info["count"]
        },
        "learning": {
            "what_happened": f"Agent struggled: {struggle_info['struggle_type']}",
            "pattern": struggle_info["struggle_type"],
            "improvement": "Detect earlier and try different approach"
        }
    }

    filepath = diary_dir / f"struggle-{entry['id']}.json"
    with open(filepath, 'w') as f:
        json.dump(entry, f, indent=2)


def main():
    try:
        input_data = json.load(sys.stdin)
    except Exception:
        print(json.dumps({"permission": "allow"}))
        sys.exit(0)

    command = input_data.get("command", "")
    conversation_id = input_data.get("conversation_id", "unknown")

    if not command:
        print(json.dumps({"permission": "allow"}))
        sys.exit(0)

    # Load history
    history = load_history(conversation_id)

    # Detect struggling
    struggle_info = detect_struggling(command, history)

    # Update history with this command
    history = update_history(history, command)
    save_history(conversation_id, history)

    # Build output
    output = {"permission": "allow"}

    if struggle_info["is_struggling"]:
        # Log to diary
        log_struggle(conversation_id, command, struggle_info)

        # Inject guidance
        output["agent_message"] = f"[Shipmate] {struggle_info['suggestion']}"

    print(json.dumps(output))
    sys.exit(0)


if __name__ == "__main__":
    main()
